File: python/ValidNumber.py
class Solution:
    # @param s, a string
    # @return a boolean
    def isInteger(self, s) :
        return s.isdigit()
    def isSignInteger(self, s):
        return s!='' and s[0] in ['+', '-'] and self.isInteger(s[1:])
    def validInteger(self, s) :
        return self.isInteger(s) or self.isSignInteger(s)
    def isPointNumber(self, s) :
        v = s.split('.')
        return len(v) == 2 and (v[0] != '' or v[1] != '') and  \
            (v[1] == '' or self.isInteger(v[1])) and \
            (v[0] == '' or self.isInteger(v[0]))
    def isSignPointNumber(self, s) :
        return s != '' and s[0] in ['+', '-'] and self.isPointNumber(s[1:])
    def validPointNumber(self, s) :
        return self.isPointNumber(s) or self.isSignPointNumber(s)
    def validNumber(self, s) :
        return self.validInteger(s) or self.validPointNumber(s)
    def isNumber(self, s):
        v = s.strip().split('e')
        eNum = '' if len(v) < 2 else v[1]
        return len(v) <= 2 and (self.validNumber(v[0]) and (self.validInteger(eNum) if 'e' in s else True))


sol = Solution()
print sol.isNumber('0')
print sol.isNumber(' 0.1')
print sol.isNumber('abc')
print sol.isNumber('1 a')
print sol.isNumber('2e10')
print sol.isNumber('0e')

File: cpluplus/ValidNumber.cpp
class Solution {
public:
    bool isPosInteger(const char* s, int l, int r)
    {
        if(l > r) return false;
        for(int i = l; i <= r; i++)
        {
            if(*(s + i) < '0' || *(s + i) > '9') return false;
        }
        return true;
    }
    bool isInteger(const char* s, int l, int r)
    {
        if(*(s + l) == '-' || *(s + l) == '+')
        {
            return isPosInteger(s, l + 1, r);
        }
        return isPosInteger(s, l, r);
    }
    bool isPosFloat(const char* s, int l, int r)
    {
        int il = l;
        int ir;
        int dl;
        int dr;
        for(int i = l; i <= r; i++)
        {
            if(*(s + i) == '.')
            {
                ir = i - 1;
                dl = i + 1;
                dr = r;
                if(il > ir && dl > dr) return false;
                return (il > ir || isPosInteger(s, il, ir)) && 
                       (dl > dr || isPosInteger(s, dl, dr));
            }
        }
        return isPosInteger(s, il, r);
    }
    bool isFloat(const char* s, int l, int r)
    {
        if(*(s + l) == '-' || *(s + l) == '+')
        {
            return isPosFloat(s, l + 1, r);
        }
        return isPosFloat(s, l, r);
    }
    bool isNumber(const char *s) {
        int l = 0;
        while(*(s + l) == ' ')
        {
            l++;
        }
        int r = strlen(s) - 1;
        while(*(s + r) == ' ')
        {
            r--;
        }
        int il = l;
        int ir;
        int dl;
        int dr;
        for(int i = l; i <= r; i++)
        {
            if(*(s + i) == 'e')
            {
                ir = i - 1;
                dl = i + 1;
                dr = r;
                return isFloat(s, il, ir) && isInteger(s, dl, dr);
            }
        }
        return isFloat(s, il, r);
    }
};
